### DISCUSSION TITLE
Where are my uploaded data and its metadata on the volume

### DISCUSSION THREAD

**Discussion by lyudmilalala** [seaweedfs/seaweedfs#4382] [Q&A]
I am new to SeaweedFS, and I am trying to understand how SeaweedFS are storing my data and metadata. I am afraid if I do not make things clear now, I will have trouble for data migrations later.

It would be nice if someone could tell me what will happen step by step when an upload request is send to the s3 server. I did not find workflow like this in the wiki docs.

Here are some experiments I have done.

After I run `weed server -dir=./data -s3` on local host 192.168.0.92, I find out the following files under my `./data` directory:
```
$ tree data/
data/
├── filerldb2
│   ├── 00     # these are the same for each volume directory, so I only write once here
│   │   ├── 000001.log
│   │   ├── CURRENT
│   │   ├── LOCK
│   │   ├── LOG
│   │   └── MANIFEST-000000
│   ├── 01
│   ├── 02
│   ├── 03
│   ├── 04
│   ├── 05
│   ├── 06
│   └── 07
├── m9333
│   ├── conf
│   ├── log
│   └── snapshot
└── vol_dir.uuid
```

If I create an s3 bucket and upload a file by the following:
```
s3 = boto3.client('s3',endpoint_url='http://192.168.0.92:8333',aws_access_key_id='123',aws_secret_access_key='456')

# Create a new bucket
bucket_name = 'demo'
s3.create_bucket(Bucket=bucket_name)
# Upload a file to the bucket
with open('data/test.txt', 'rb') as data:
    s3.upload_fileobj(data, bucket_name, 'test.txt')
```

Files for 8 volumes has been created, and I find some bucket information in `8.dat`, and content of my uploaded data in `demo_1.dat`. Later, when I upload some files to the bucket, `8.dat` will be updated and content of the files seems to be put in one of `demo_*.dat`.
 
```
$ tree data/
data/
├── 8.dat
├── 8.idx
├── 8.vif
├── demo_1.dat
├── demo_1.idx
├── demo_1.vif
├── demo_2.dat
├── demo_2.idx
├── demo_2.vif
├── demo_3.dat
├── demo_3.idx
├── demo_3.vif
├── demo_4.dat
├── demo_4.idx
├── demo_4.vif
├── demo_5.dat
├── demo_5.idx
├── demo_5.vif
├── demo_6.dat
├── demo_6.idx
├── demo_6.vif
├── demo_7.dat
├── demo_7.idx
├── demo_7.vif
├── filerldb2
│   ├── 00
│   │   ├── 000001.log
│   │   ├── CURRENT
│   │   ├── LOCK
│   │   ├── LOG
│   │   └── MANIFEST-000000
│   ├── 01
│   ├── 02
│   ├── 03
│   ├── 04
│   ├── 05
│   ├── 06
│   └── 07
├── m9333
│   ├── conf
│   ├── log
│   └── snapshot
└── vol_dir.uuid
```

When I create a second bucket, new volume files are created, but no new folders are created under `filerldb2`.

```
$ tree data/
data/
├── 8.dat
├── 8.idx
├── 8.vif
├── demo_1.dat
├── demo_1.idx
├── demo_1.vif
├── demo2_10.dat
├── demo2_10.idx
├── demo2_10.vif
├── demo2_11.dat
├── demo2_11.idx
├── demo2_11.vif
├── demo2_12.dat
├── demo2_12.idx
├── demo2_12.vif
├── demo2_13.dat
├── demo2_13.idx
├── demo2_13.vif
├── demo2_14.dat
├── demo2_14.idx
├── demo2_14.vif
├── demo2_15.dat
├── demo2_15.idx
├── demo2_15.vif
├── demo2_9.dat
├── demo2_9.idx
├── demo2_9.vif
├── demo_2.dat
├── demo_2.idx
├── demo_2.vif
├── demo_3.dat
├── demo_3.idx
├── demo_3.vif
├── demo_4.dat
├── demo_4.idx
├── demo_4.vif
├── demo_5.dat
├── demo_5.idx
├── demo_5.vif
├── demo_6.dat
├── demo_6.idx
├── demo_6.vif
├── demo_7.dat
├── demo_7.idx
├── demo_7.vif
├── filerldb2
│   ├── 00
│   │   ├── 000001.log
│   │   ├── CURRENT
│   │   ├── LOCK
│   │   ├── LOG
│   │   └── MANIFEST-000000
│   ├── 01
│   ├── 02
│   ├── 03
│   ├── 04
│   ├── 05
│   ├── 06
│   └── 07
├── m9333
│   ├── conf
│   ├── log
│   └── snapshot
└── vol_dir.uuid
```

Base on the experiment results, I have a few questions:
1. Why there is an 8th volume? It seems as a manager. Is it means when I setup N volumes on a server, I need to prepare disk enough for N+1 volumes, because a volume used to store file metadata will be created there? Is the 8th volume related with the filer?
2. What are each  `demo_*.idx`, `demo_*.dat`,  `demo_*.vif` used for? Are they related with the folders under `filerldb2`? If I want to move part of the volume to a new server, is it enough for me to move these three files there?
3. Are all data stored in a volume be combined into one `.dat` file. Is there a way for me to get the content of a specific file I uploaded to the bucket (for backup or check something)? I know that large file will be chunked and put into different volumes. So I feel it is even harder to get back the data once one of the weed servers (master, volume, s3) is broken.
4. Why no new folders are created under `filerldb2` after a second bucket has been created?
5. In `8.dat` I see `/buckets/demo2` in the file info line. Is it a replication of the bucket? If I have another self-created bucket called `demo2`, will they affect each other?
6. 
![image](https://user-images.githubusercontent.com/32922504/230304118-e90403b8-3a51-493f-9735-eff79ab5ce6c.png)

Also, if I visit the filer at http://192.168.0.92:8888/ after I have already created the buckets, no extra volumes will be created. But if I visit the filer before creating any bucket, it will create some default volume files as below. Why does this happen?
```
$ ls ./data/
1.dat  1.idx  1.vif  2.dat  2.idx  2.vif  3.dat  3.idx  3.vif  4.dat  4.idx  4.vif  5.dat  5.idx  5.vif  6.dat  6.idx  6.vif  7.dat  7.idx  7.vif  filerldb2  m9333  vol_dir.uuid
```

**Comment by rernst76** [comment]
> Why there is an 8th volume? It seems as a manager. Is it means when I setup N volumes on a server, I need to prepare disk enough for N+1 volumes, because a volume used to store file metadata will be created there? Is the 8th volume related with the filer?

Unless you specified a `max` value when starting the volume server, you will always get 8 volumes. The default volume size is 30 GB and default number of volumes is 8. So you should _always_ ensure that you have `volumeSizeLimitMB` * `max` space wherever you are defining your data directory.

> What are each demo_*.idx, demo_*.dat, demo_*.if used for? Are they related with the folders under filerldb2? If I want to move part of the volume to a new server, is it enough for me to move these three files there?

I personally would not be moving anything around manually. You can use weed shell for this if you want to move live volumes. See: https://github.com/seaweedfs/seaweedfs/wiki/weed-shell

> Are all data stored in a volume be combined into one .dat file. Is there a way for me to get the content of a specific file I uploaded to the bucket (for backup or check something)? I know that large file will be chunked and put into different volumes. So I feel it is even harder to get back the data once one of the weed servers (master, volume, s3) is broken.

There is likely a way to do this, but it is beyond my knowledge. You should always be able to move the bucket and start another master/volume server. Then you can use either weed shell or s3.

> Why no new folders are created under filerldb2 after a second bucket has been created?

`filerldbd2` is the ldb2 directory that the Filer server is using. I don't think you can draw any conclusions about how ldb2 is being used here. It is just the database that is used by Filer.

> In 8.dat I see /buckets/demo2 in the file info line. Is it a replication of the bucket? If I have another self-created bucket called demo2, will they affect each other?

The .dat file is the store file where the actual files are stored. Seaweed follows a similar design to [Haystack](https://engineering.fb.com/2009/04/30/core-infra/needle-in-a-haystack-efficient-storage-of-billions-of-photos/). 

`/buckets/demo2` and `demo2` are two separate paths. I would not expect any conflicts -- but I am not positive. If this is important you should test it. The conflict, if it does occur, would likely happen with `filer`, as it is the component that maps fid<-> paths. Master and volume servers only concern themselves with fids.

**Comment by lyudmilalala** [comment]
@rernst76 Thanks for your answer.

So if I have multiple folders/buckets of files in my local test environment, and now I want to move only part of them to a different seaweedfs environment not in the same network. What is the most efficient way to do the thing?

If I would like to migrate by moving volume directories directly to another computer, I also need to move the filerldbd2, right? Otherwise, I cannot access the files properly by filer api? (I test this case and come up with this conclusion. Correct me if I am wrong.)

To my understanding, it is hard to directly migrating volume directories, because we cannot filter out the subdirectories we need easily. As a result, the only other way of migration comes up to my mind is calling filer api to read from environment A and write to environment B. However I feel it's really inefficient. Is there a better way?

**Reply by rernst76** [reply]
This is a tough one. Maybe someone more of an expert has a better answer, but I will give it a shot anyway.

As you state, I don't think you will be able to directly move volumes without having some filer treatment as well, at least while maintaining metadata (paths/directories).

So here is what I would try.

- Move volume directories manually via scp, sneakernet, etc. to another volume server
- Use weed shell to export filer metadata via: ` weed fs.meta.save`
- On the new filer load the exported filer metadata via: `weed fs.meta.load`

I have never tried this, so have no idea if it works, but it is where I would start.


If you try it out, let me know how it works!

P.S.: I tried to honor the constraint you suggested with "not in the same network", but if there is any flexibility with this, one of the tools I often reach for in this kind of scenario is wireguard, usually via Tailscale, which would allow you to create an overlay network that would securely connect the two machines. Obviously, this might not be an option in some environments, but I figured I would mention it anyway.
